homeassistant/components/modbus/__init__.py

```diff
@@ -253,6 +257,13 @@
             vol.Optional(CONF_STEP, default=0.5): vol.Coerce(float),
             vol.Optional(CONF_TEMPERATURE_UNIT, default=DEFAULT_TEMP_UNIT): cv.string,
             vol.Optional(CONF_HVAC_ONOFF_REGISTER): cv.positive_int,
+            vol.Optional(CONF_HVAC_ONOFF_REGISTER_WRITE_TYPE): vol.In(
+                [
+                    CALL_TYPE_WRITE_COIL,
+                    CALL_TYPE_WRITE_REGISTER,
+                    CALL_TYPE_WRITE_REGISTERS,
```

Comment:
Since this is intended to overwrite whatever WRITE_REGISTERS dictates but also should not break existing installations I honestly can't see how this optional setting, without a default, would conflict.

My only alternative would be using/creating CONF_HVAC_ONOFF_REGISTER_USE_COIL instead (And potentially also CONF_HVAC_ONOFF_REGISTER_USE_COILS) this would however make the code more complicated

Reply:
WRITE_REGISTERS does basically the same as you do with TYPE_REGISTERS/S so there are a clear conflict, like what happens if a configuration have both, which parameter wins...

We will not allow such conflicts, just look how it's done elsewhere.

Reply:
Ok, let me explain

The climate entity:
* Assumes it can be fully controlled by reading/writing register(s) (not coils!)
* Assumes that all hvac modes can be controlled by reading/writing register(s) (not coils!)
* Allows the user to configure a different register (not coil!) address when the "on/of" state is not part of the hvac address

This new s setting (like the name suggests) adds the option to use a coil instead of register(s), **only for the on/off state**

The on/off state is, sort of, part of the hvac mode setting, but it's handled differently
As written, 

* WRITE_REGISTERS will still keep working as intended for setting the hvac modes
* WRITE_REGISTERS will only NOT be checked when the user configured the optional CONF_HVAC_ONOFF_REGISTER_WRITE_TYPE setting.

There will always be room for misconfiguration, but I see no other way that would either need way more if statements/complexity, or backwards incompatible changes. Both I really have no desire for

To answer your question
> like what happens if a configuration have both, which parameter wins...

For hvac mode, there's not even a competition, since CONF_HVAC_ONOFF_REGISTER_WRITE_TYPE  is not used in that logic

For on/off state, CONF_HVAC_ONOFF_REGISTER_WRITE_TYPE is of course the winner
The whole point is that we need a coil, and not register(s)

I'm very open to suggestions, but I'm also very sure that there is no "conflict" here